route rule: ensure idempotency for USE_DEFAULT_PRIORITY

ihuguet opened the pull request:
When the desired state has route rules without a priority defined,
nmstate creates them with increasing values of priority. However,
this was causing that applying the same desired state twice created
the rule twice with different priorities. As a result, the operation
was not idempotent.

Fix it by searching for a matching rule in the current state and, if
there is any, use its priority value to fill the priority value of the
desired rule if it's empty. As a result, new rules are not created if
one that matches already exists.

Fix also a helper function of the integration tests to make it asserting
that the rule is not duplicated.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

ihuguet committed:
route rule: avoid adding twice to merged state

Routing roules defined in the desired state with `state: present` were
being added twice to the merged state if they already existed in the
current state.

Fix it by adding them to the merged state only if they are actually new
rules.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

ihuguet committed:
test: fix ip_rule_exist_in_os to properly assert rules with src=all

Prior to this change, any kernel rule with src=all or dst=all was
skipped, so asserting rules that actually had “all” never failed and
regressions slipped through. Now we normalize and check every rule
before matching, so tests for src=all/dst=all will correctly pass or fail.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>

liangwen12year commented on tests/integration/testlib/iprule.py:
```diff
@@ -37,24 +37,26 @@ def ip_rule_exist_in_os(rule):
-        if ip_to and ip_to != iplib.to_ip_address_full(
-            rule["dst"], rule.get("dstlen")
-        ):
+        if ip_to is not None and rule.get("dst") != ip_to:
```
Nit, this patch's commit message is a bit hard to understand, maybe rephrase to: 

```
Prior to this change, any kernel rule with src=all or dst=all was
skipped, so asserting rules that actually had “all” never failed and
regressions slipped through. Now we normalize and check every rule
before matching, so tests for src=all/dst=all will correctly pass or fail.
```

ihuguet replied:
Done

ihuguet committed:
route rule: ensure idempotency for USE_DEFAULT_PRIORITY

When the desired state has route rules without a priority defined,
nmstate creates them with increasing values of priority. However,
this was causing that applying the same desired state twice created
the rule twice with different priorities. As a result, the operation
was not idempotent.

Fix it by searching for a matching rule in the current state and, if
there is any, use its priority value to fill the priority value of the
desired rule if it's empty. As a result, new rules are not created if
one that matches already exists.

Fix also a helper function of the integration tests to make it asserting
that the rule is not duplicated.

Signed-off-by: Íñigo Huguet <ihuguet@redhat.com>
